internal/configs/virtualserver.go

```diff
@@ -264,20 +264,30 @@ func generateProxyPassProtocol(upstream conf_v1alpha1.Upstream) string {
 	return "http"
 }
 
+func generateStringOrDefault(s string, defaultS string) string {
+	if s == "" {
+		return defaultS
+	}
+	return s
+}
+
 func generateLocation(path string, upstreamName string, upstream conf_v1alpha1.Upstream, cfgParams *ConfigParams) version2.Location {
 	return version2.Location{
-		Path:                 path,
-		Snippets:             cfgParams.LocationSnippets,
-		ProxyConnectTimeout:  generateTime(upstream.ProxyConnectTimeout, cfgParams.ProxyConnectTimeout),
-		ProxyReadTimeout:     generateTime(upstream.ProxyReadTimeout, cfgParams.ProxyReadTimeout),
-		ProxySendTimeout:     generateTime(upstream.ProxySendTimeout, cfgParams.ProxySendTimeout),
-		ClientMaxBodySize:    cfgParams.ClientMaxBodySize,
-		ProxyMaxTempFileSize: cfgParams.ProxyMaxTempFileSize,
-		ProxyBuffering:       cfgParams.ProxyBuffering,
-		ProxyBuffers:         cfgParams.ProxyBuffers,
-		ProxyBufferSize:      cfgParams.ProxyBufferSize,
-		ProxyPass:            fmt.Sprintf("%v://%v", generateProxyPassProtocol(upstream), upstreamName),
-		HasKeepalive:         upstreamHasKeepalive(upstream, cfgParams),
+		Path:                     path,
+		Snippets:                 cfgParams.LocationSnippets,
+		ProxyConnectTimeout:      generateTime(upstream.ProxyConnectTimeout, cfgParams.ProxyConnectTimeout),
+		ProxyReadTimeout:         generateTime(upstream.ProxyReadTimeout, cfgParams.ProxyReadTimeout),
+		ProxySendTimeout:         generateTime(upstream.ProxySendTimeout, cfgParams.ProxySendTimeout),
+		ClientMaxBodySize:        cfgParams.ClientMaxBodySize,
+		ProxyMaxTempFileSize:     cfgParams.ProxyMaxTempFileSize,
+		ProxyBuffering:           cfgParams.ProxyBuffering,
+		ProxyBuffers:             cfgParams.ProxyBuffers,
+		ProxyBufferSize:          cfgParams.ProxyBufferSize,
+		ProxyPass:                fmt.Sprintf("%v://%v", generateProxyPassProtocol(upstream), upstreamName),
+		ProxyNextUpstream:        generateStringOrDefault(upstream.ProxyNextUpstream, "error timeout"),
+		ProxyNextUpstreamTimeout: generateStringOrDefault(upstream.ProxyNextUpstreamTimeout, "0s"),
```

Comment:
There is a function called `generateTime` that does pretty much the same. I think it might be a good time to remove `generateTime` and use `generateString` for times and strings, what do you think? They do pretty much the same but with a different name.

Reply:
@Rulox good suggestion 👍